packages/eslint-plugin/src/rules/no-deprecated-components.ts

```diff
@@ -0,0 +1,167 @@
+/* !
+ * (c) Copyright 2022 Palantir Technologies Inc. All rights reserved.
+ */
+
+import { TSESTree } from "@typescript-eslint/utils";
+
+import { createRule } from "./utils/createRule";
+
+const DEPRECATED_TO_NEW_MAPPING: { [deprecated: string]: string } = {
+    AbstractComponent: "AbstractComponent2",
+    AbstractPureComponent: "AbstractPureComponent2",
+    CollapsibleList: "OverflowList",
+    Popover: "Popover2",
+    Tooltip: "Tooltip2",
+};
+
+type MessageIds = "nonDeprecated" | "deprecated";
+
+// tslint:disable object-literal-sort-keys
+export const noDeprecatedComponentsRule = createRule<unknown[], MessageIds>({
+    name: "no-deprecated-components",
+    meta: {
+        type: "suggestion",
+        docs: {
+            description: "Recommends using non-deprecated versions of the component or constant instead",
+            requiresTypeChecking: false,
+            recommended: "error",
+        },
+        messages: {
+            nonDeprecated: "Replace deprecated component {{ deprecated }} with {{ nonDeprecated }}",
+            deprecated: "This component or constant is deprecated with no replacement.",
+        },
+        schema: [
+            {
+                enum: ["nonDeprecated", "deprecated"],
+            },
+        ],
+    },
+    defaultOptions: [],
+    create: context => {
+        const deprecatedImports: Array<
+            | { namespace: string; type: "namespace" }
+            | { type: "function"; functionName: string; localFunctionName: string }
+        > = [];
+
+        function isDeprecatedComponent(name: string) {
+            if (deprecatedImports.length === 0) {
+                return false;
+            }
+
+            return deprecatedImports.some(
+                deprecatedImport =>
+                    (deprecatedImport.type === "function" && deprecatedImport.localFunctionName === name) ||
+                    DEPRECATED_TO_NEW_MAPPING[name] != null,
+            );
+        }
+
+        function isDeprecatedNamespacedComponent(name: string, property: string) {
+            return (
+                deprecatedImports.some(
+                    deprecatedImport => deprecatedImport.type === "namespace" && deprecatedImport.namespace === name,
+                ) && DEPRECATED_TO_NEW_MAPPING[property] != null
+            );
+        }
+
+        // Get the list of all deprecated imports from blueprint
+        return {
+            "Program > ImportDeclaration": (node: TSESTree.ImportDeclaration) => {
+                if (node.source.value !== "@blueprintjs/core") {
```

Comment:
This needs to check for other packages as well, like `@blueprintjs/select`. We could maintain a list of packages which have deprecated exports at the top of this file.